Add documentation for using DTOs in form handling

ckrack opened the pull request:
Add a short article about using data transfer objects in form handling.
The docs use a new make:dto command.
Merge when [make:dto](https://github.com/symfony/maker-bundle/pull/303) is merged.

ckrack committed:
Add docs for using DTOs in forms

ckrack committed:
Add note and link to forms docs

ckrack committed:
Correct mistakes

Use correct generated code for the DTO.
Fix diff indenting.
Specify DateType field.

yceruto commented on form/data_transfer_objects.rst:
```diff
@@ -0,0 +1,358 @@
+How to use Data Transfer Objects (DTO)
+======================================
+
+Data transfer objects can be used by forms to separate entitites from the
```
`entities`

yceruto commented on form/data_transfer_objects.rst:
```diff
@@ -0,0 +1,358 @@
+~~~~~~~~~~~~
+
+In applications using :doc:`Symfony Flex </setup/flex>`, run this command to
+install the maker bundle before using it:
```
`Maker`

yceruto commented on form/data_transfer_objects.rst:
```diff
@@ -0,0 +1,358 @@
+
+    $ composer require maker --dev
+
+You will also need these packages in order to proceed with creating a crud
```
`CRUD`

yceruto commented on form/data_transfer_objects.rst:
```diff
@@ -0,0 +1,358 @@
+.. index::
+   single: Creating a data transfer object
+
+Creating a data transfer object
```
`Data Transfer Object`?

yceruto commented on form/data_transfer_objects.rst:
```diff
@@ -0,0 +1,358 @@
+    use App\Entity\Task;
+    + use App\Form\Data\TaskData;
+
+    ...
```
`// ...`

yceruto commented on form/data_transfer_objects.rst:
```diff
@@ -0,0 +1,358 @@
+    class TaskController extends AbstractController
+    {
+
+    ...
```
Same here

yceruto commented on forms.rst:
```diff
@@ -695,6 +695,14 @@ the choice is ultimately up to you.
     to modify it, use the :method:`Symfony\\Component\\Form\\FormFactoryInterface::createNamed` method.
     You can even suppress the name completely by setting it to an empty string.
 
+Using data transfer objects
```
`... Data Transfer Objects`

ckrack committed:
Corrections after review

Fix naming of Maker and Data Transfer Objects.
Replace ... placeholders with comments.

ckrack reviewed (commented):
Thank you @yceruto!
Updated with the changes.

ckrack committed:
Fix too short title underline

ckrack committed:
Remove an outdated part of a dialogue

ckrack committed:
Fix typos

ckrack committed:
Use more advanced example Task entity

The diff is against the Task entity that
would originate from following
the forms main library including the validation section.

ckrack committed:
Remove explicitely set DateTimeType

This was replaced by using @Assert\Type(DateTime) annotation.
Related to: https://github.com/symfony/symfony/issues/29029

ckrack commented:
Build fails due to this commit @javiereguiluz https://github.com/symfony/symfony-docs/commit/c35b930687874cf9140b6936ddfc41ed04d06530#diff-f7bfae99d3f5b1553a7bb226b57d1d91

kunicmarko20 commented on form/data_transfer_objects.rst:
```diff
@@ -0,0 +1,347 @@
+When entities are used as data classes for a form, the data is injected into
+the entity and validated.
+When the validation fails, the invalid data is still left in the entity.
+This can lead to invalid data being saved in the database.
```
Or maybe unexpected errors? if data is invalid it probably won't even get to be saved in the database.

ckrack replied:
It can easily end up in the database.
Example: Email validation fails, entity is saved in a listener or somewhere further down, database field expects a string, so it won't fail on the insert.

kunicmarko20 replied:
Good point, but it can also lead to exceptions. So maybe just add `or unexpected exceptions`?

kunicmarko20 commented on form/data_transfer_objects.rst:
```diff
@@ -0,0 +1,347 @@
+When the validation fails, the invalid data is still left in the entity.
+This can lead to invalid data being saved in the database.
+
+You will use the Maker bundle to highlight the differences between using DTOs
```
`You` should maybe be `We`?

ckrack replied:
In [forms.rst](https://github.com/symfony/symfony-docs/blob/master/forms.rst), you is being used.
I think it's the better choice, as it sounds more empowering. The reader builds, we just teach them how.

nicolas-grekas replied:
this sentence sounds strange, "You will etc"...: What? No I won't! :)
Maybe: 
Let's use the Maker bundle to highlight the differences between using DTOs and entities.

kunicmarko20 commented on form/data_transfer_objects.rst:
```diff
@@ -0,0 +1,347 @@
+
+Use the example ``Task`` entity from :doc:`the main forms tutorial </forms>`,
+make it a doctrine entity (this requires adding a primary key ``id``) and make
+the setters fluent.
```
Do they have to be fluent? Not really a fan of it, maybe check out https://ocramius.github.io/blog/fluent-interfaces-are-evil/

ckrack replied:
It's a valid point, although it should probably be discussed [in the PR](https://github.com/symfony/maker-bundle/pull/303).
In `make:entity` the code is generated with fluent setters and I think the generated code in the `Maker` bundle should resemble itself.
I'm probably still going to change it, as `make:dto` is supposed to work with existing entities and I can't assume that they have fluent interfaces.

kunicmarko20 replied:
Since it is already in the Maker I guess it makes sense then

ckrack replied:
I have updated this.
The example is based on the example entity from the forms docs. As there are no fluent setters, we can just omit this change.
And `make:dto` must work without fluent setters anyways - thanks for pointing me onto that.

kunicmarko20 replied:
Glad I helped. :smile:

kunicmarko20 commented on form/data_transfer_objects.rst:
```diff
@@ -0,0 +1,347 @@
+    +    * @ORM\GeneratedValue()
+    +    * @ORM\Column(type="integer")
+    +    */
+    +   protected $id;
```
why are properties protected?

ckrack replied:
See [forms.rst#creating-a-simple-form](https://github.com/symfony/symfony-docs/blob/master/forms.rst#creating-a-simple-form)
I'm iterating on this example.

[Further down on the same page, one property is even public](https://github.com/symfony/symfony-docs/blob/master/forms.rst#form-validation).
Fix both in a separate PR?

kunicmarko20 replied:
I would fix both, no need for those fields to be protected.

kunicmarko20 commented on form/data_transfer_objects.rst:
```diff
@@ -0,0 +1,347 @@
+Notice that it uses the ``Task`` entity by default.
+This means that the form data is injected into the ``Task`` entity directly and validated with the annotations.
+
+Replace this with ``TaskData`` to prevent the aforementioned problems with an invalid entity.
```
`aforementioned` => `before mentioned`?

ckrack replied:
I think aforementioned is a nice word and perfectly suitable in this case, as it would have to be `as mentioned before` instead, which is much longer.

kunicmarko20 replied:
Hm, I can agree with that.

kunicmarko20 commented on form/data_transfer_objects.rst:
```diff
@@ -0,0 +1,347 @@
+
+          if ($form->isSubmitted() && $form->isValid()) {
+    +         $task = $taskData->fill(new Task());
+              $em = $this->getDoctrine()->getManager();
```
please don't abbreviate, write `$entityManager`

ckrack replied:
I agree, however this should be fixed in [make:crud](https://github.com/symfony/maker-bundle/blob/master/src/Resources/skeleton/crud/controller/Controller.tpl.php#L49).

ckrack replied:
Updated, assuming the change will be merged.

kunicmarko20 commented on form/data_transfer_objects.rst:
```diff
@@ -0,0 +1,347 @@
+        *
+        * @param Task|null $task
+        */
+        public function __construct(? Task $task = null)
```
is the space after `?` intentional?

kunicmarko20 commented on form/data_transfer_objects.rst:
```diff
@@ -0,0 +1,347 @@
+        /**
+        * Create DTO, optionally extracting data from a model.
+        *
+        * @param Task|null $task
```
This looks redundant to me

ckrack replied:
Correct, I've already changed this in the generated code - updated the example.

ckrack committed:
Update generated DTO

ckrack committed:
Use private properties

The example contained protected and public properties on the entity.
Replaced with private.

ckrack committed:
Remove abbreviated var name

Assumes that make:crud generates code without abbreviation.

ckrack committed:
Remove fluent interface

The example entity does not use a fluent interface.
This also makes the generated code easier to read.

ckrack committed:
Mention unexpected exceptions

ckrack committed:
Fix generated code

The type is specified.

nicolas-grekas commented on form/data_transfer_objects.rst:
```diff
@@ -0,0 +1,316 @@
+
+Data Transfer Objects can be used by forms to separate entities from the
+validation logic of forms.
+Entities should always have a valid state.
```
should be removed - adds no value IMHO

nicolas-grekas commented on form/data_transfer_objects.rst:
```diff
@@ -0,0 +1,316 @@
+validation logic of forms.
+Entities should always have a valid state.
+When entities are used as data classes for a form, the data is injected into
+the entity and validated.
```
When entities are used as data classes for a form, validation happens after the form data has been injected into the entities.

nicolas-grekas commented on form/data_transfer_objects.rst:
```diff
@@ -0,0 +1,316 @@
+Entities should always have a valid state.
+When entities are used as data classes for a form, the data is injected into
+the entity and validated.
+When the validation fails, the invalid data is still left in the entity.
```
in the entities.

nicolas-grekas commented on form/data_transfer_objects.rst:
```diff
@@ -0,0 +1,316 @@
+    $ composer require form validator twig-bundle orm-pack security-csrf annotations
+
+Use the example ``Task`` entity from :doc:`the main forms tutorial </forms>`
+and make it a doctrine entity (this requires adding a primary key ``id``).
```
Doctrine

nicolas-grekas commented on form/data_transfer_objects.rst:
```diff
@@ -0,0 +1,316 @@
+    namespace App\Entity;
+
+    + use Symfony\Component\Validator\Constraints as Assert;
+    + use Doctrine\ORM\Mapping as ORM;
```
alpha order for use statements?

ckrack committed:
Improve wording and example

webda2l commented on form/data_transfer_objects.rst:
```diff
@@ -0,0 +1,316 @@
+        public function fill(Task $task): Task
+        {
+            $task->setTask($this->task);
+            $task->setDueDate($this->dueDate)
```
```suggestion
            $task->setDueDate($this->dueDate);
```

noniagriconomie commented on forms.rst:
```diff
@@ -695,6 +695,14 @@ the choice is ultimately up to you.
+Using Data Transfer Objects
+---------------------------
+
+There are some problems when using entities as directly mapped data classes for forms.
```
Maybe not the word problems, but notice or info 
Better write something like :

Alternatively, you can decouple your entities creation with DTO that are bound to form instead, see more
Etc

nicolas-grekas commented:
This PR has been closed because the master has been removed.
Please submit it again against the appropriate branch.
